flink-contrib/docker-flink/Dockerfile

```diff
@@ -36,22 +31,24 @@ ENV PATH $PATH:$FLINK_HOME/bin
 EXPOSE 8081
 EXPOSE 6123
 
+# flink-dist can point to a directory or a tarball on the local system
+ARG flink_dist=NOT_SET
+
 # Install build dependencies and flink
+ADD $flink_dist $FLINK_INSTALL_PATH
 RUN set -x && \
-  mkdir -p $FLINK_INSTALL_PATH && \
-  apk --update add --virtual build-dependencies curl && \
-  curl -s $(curl -s https://www.apache.org/dyn/closer.cgi\?preferred\=true)flink/flink-${FLINK_VERSION}/flink-${FLINK_VERSION}-bin-hadoop${HADOOP_VERSION}-scala_${SCALA_VERSION}.tgz | \
```

Comment:
Hi guys, thanks for picking up this PR and running with it.  I think we should separate the "official" Docker images issue from this PR.  This PR was about just simply improving the existing Docker tooling and making it easier for Flink developers to build their own images, maybe with their own tweaks, or specifying specific Flink releases to build the image from.

On the "official" images topic I'm totally fine with either approach.  dA could do this and maintain those official images or we could do it as part of Flink and the community can maintain them.  The main thing for me is just seeing that it's super easy for people to get Flink running with Docker and there is some accepted "official" image for people to grab rather than wading through a bunch of Flink images generated by various people.  That's bad for the user experience.

Regardless, let's figure out separately how best to get an official image on DockerHub and get this change finished up because I think the changes here are useful as is.